### ISSUE TITLE
Band math operator misuse

### ISSUE THREAD

**Issue by soxofaan** [Open-EO/openeo-python-client#123]
The Python client provides intuitive expression of "band math" with mathematical operators (e.g. `cube.band("red") - 0.8 * cube.band("blue")` and translates this behind the scenes to a callback in a reduce process.

Current implementation of these operators assume that the user correctly picked bands with the `.band()` and that the user intends to do reduction along spectral/band dimension.

However, it is pretty easy to  (unintentionally) abuse these operators and end up with invalid process graphs. For example:

```python
cube = conn.load_collection('CGS_SENTINEL2_RADIOMETRY_V102_001')
x = cube + 1
```

results in a graph:
```
{
    "loadcollection3": {
        "process_id": "load_collection",
        "arguments": { ... }
    },
    "reduce3": {
        "process_id": "reduce",
        "arguments": {
            "data": {
                "from_node": "loadcollection3"
            }
            "dimension": "spectral_bands",
            "reducer": {
                "callback": {
                    "sum1": {
                        "result": true,
                        "process_id": "sum",
                        "arguments": {
                            "data": [
                                {"from_argument": "data"},
                                1
                            ]
                        }
...
```

problems:
- there is a reduce process along spectral dimension while `cube + 1` does not suggest any dimension
- the reduce callback will probably not work: the `sum` process will receive 2 arguments: a list (values along the spectral dimension) and an int (1). Even if a backend would support adding a scalar to an array in some way (e.g. broadcasting), it still expected to return a single scalar as as reduced result.

**Comment by soxofaan** [comment]
In the example above, it would probably be more intuitive to use a `apply` process  (application of unary process to each pixel)

**Comment by jdries** [comment]
Using apply is not allowed if you change the number of bands:

https://open-eo.github.io/openeo-api/processreference/#apply
The resolution, cardinality and the number of dimensions are the same as for the original data cube.

Using reduce over bands is basically the agreed way to do band math.

**Comment by soxofaan** [comment]
what I mean is that if a user does something like this:

    # Load 4D cube e.g.: (x, y, bands, time)
    cube = con.load_collection("S2")
    cube = cube + 1

then the user probably expects  still a 4D cube where each pixel value is incremented with one (numpy ndarray style)


Indeed, when working in "spectral band modus" like this:

    red = cube.band("red")
    blue = cube.band("blue")
    res = red + blue

then the user expects a 3D cube (with dimension reduction along spectral dimension under the hood).

The question/problem I have at the moment with for the DataCube refactor of #117 (which includes updating that latter case of  "spectral band style band math"):  how should I handle the former case (`cube + 1`)?  
The "old" 0.4 style implementation blindly assumes spectral band reduction but that results in nonsensical process graphs.
What should be done in the 1.0 style implementation?
- raise some kind of exception indicating that operators are not meant to do that. (This is what I am doing now, and it allows to simplify the logic)
-  implement numpy ndarray style broadcasting by using the `apply` process

**Comment by jdries** [comment]
Thanks I see the problem now.
Please throw an error, the fix with 'broadcasting' should be logged as a new feature, but low priority for now as we don't have a use case yet that is blocked by it.

**Comment by soxofaan** [comment]
after fixing #116 , I stumbled on a use case with "fuzzy cloud masks" that broke related to this. It goes roughly like this:

```
# 1) binary mask
clouds =  (cube.band("classification") == 4)
# 2) fuzzy mask
fuzzy = clouds.apply_kernel(kernel)
# 3) binary mask again
mask = (fuzzy > 0.1)
```

In step 1, the  `.band()` operation is currently translated to a `reduce` along "band" dimension, which is fine. 
Step 2 adds a new `apply_kernel` process to the graph
In step 3 however, the comparison is also translated in a new `reduce` process, again along "band" dimension, which does not work because that dimension is dropped after the  `reduce` of step 1.

Step 3 should be implemented as an `apply`

**Comment by soxofaan** [comment]
Another use case that is relevant here I think:

user wants to convert (S1) data from DN to dB.
current solution is `cube = cube.apply(lambda x: 10 * x.log(base=10))`
but with proper  `apply` support for "band math" like operators/methods, this could be written as
`cube = 10 * cube.log10()`

**Comment by soxofaan** [comment]
Another use case related to bad operator support in apply-mode :

```python
    # Invert again
    erode_cube = (erode_cube == 0)
    # Now dilate the mask
    dilate_cube = erode_cube.apply_kernel(...)
    
    dilate_cube = ~(dilate_cube == 0)
```

First `(erode_cube == 0)`  works, but second with additional inversion `~(dilate_cube == 0)` fails with
```text
File ~/src/openeo/openeo-python-client/openeo/rest/datacube.py:697, in DataCube.__invert__(self)
    695 @openeo_process(process_id="not", mode="operator")
    696 def __invert__(self) -> "DataCube":
--> 697     return self._operator_unary("not")

File ~/src/openeo/openeo-python-client/openeo/rest/datacube.py:649, in DataCube._operator_unary(self, operator, **kwargs)
    647 if band_math_mode:
    648     return self._bandmath_operator_unary(operator, **kwargs)
--> 649 raise OperatorException("Unsupported unary operator {op!r} (band math mode={b})".format(
    650     op=operator, b=band_math_mode))

OperatorException: Unsupported unary operator 'not' (band math mode=False)
```

current workaround is to use much uglier 
```python
dilate_cube = dilate_cube.apply(lambda x: (x == 0).not_())
```

**Comment by soxofaan** [comment]
with a349e5b, "apply mode" now supports expressions like 
- `cube = 10 * cube.log10()`
- `dilate_cube = ~(dilate_cube == 0)`
